### DISCUSSION TITLE
Mux CORS error

### DISCUSSION THREAD

**Discussion by EhanDuan** [vercel/next.js#39742] [Help]
### Overview
Hi, I am using `nextjs`, `Mux`, `firestore` to build an app

### Problem 
I used `Mux` `directUpload` to upload a video. When `asset`(term in `Mux`) ready, I will push relative data into `firestore`.

Everything is okay, when I run locally.

But when I deployed on `Netlify`, there comes the CORS error as follows:
![image](https://user-images.githubusercontent.com/75397936/185527084-c2725f15-6879-439f-9c6a-53152e9499f4.png)

Here is what I found in `Mux` doc about [making API Request](https://docs.mux.com/guides/video/make-api-requests#cors-and-client-side-api-requests).

```
[CORS Error in Browser](https://docs.mux.com/guides/video/make-api-requests#cors-error-in-browser)
request has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
>This is expected. Although making API requests directly from the browser or your mobile app would be convenient, it leaves a massive security hole in your application by the fact that your client side code would contain your API keys. Anyone who accesses your application would have the ability to steal your API credentials and make requests to Mux on your behalf. An attacker would be able to gain full control of your Mux account.

>Mux API Credentials should never be stored in a client application. All Mux API calls should be made from a trusted server.

> Instead of trying to make API requests from the client, the flow that your application should follow is:

>Client makes a request to your server
Your server makes an authenticated API request to Mux
Your server saves whatever it needs in your database
Your server responds to the client with only the information that the client needs. For example, with live streaming that's the stream key for a specific stream, for uploads that's just the direct upload URL

I followed the instruction, creating upload and push video data in firestore are 2 requests from the client side.

 I also set CORS is '*' in order to match all websites ( although insecure) to check if it is CORS value error. It was still not working. 

Thanks for your help!

### Code
`/pages/mux_upload` （part of code)

```js
const createUpload = async () => {
    try {
      console.log("entering createUpload");
      return fetch("/api/upload", {
        method: "POST"
      })
        .then((res) => res.json())
        .then(({ id, url }) => {
          console.log(`setting uploadId: ${id}`);
          setUploadId(id);
          return url;
        });
    } catch (e) {
      console.error("Error in createUpload", e);
      setErrorMessage("Error creating upload");
    }
  };

const startUpload = (evt) => {
    if (title === "") {
      alert("Video title cannot be empty");
      return;
    }

    setIsUploading(true);
    console.log("after title check:" + uploadId);
    // Returns an instance of UpChunk and begins uploading the specified File.
    const upChunk = UpChunk.createUpload({
      endpoint: createUpload,
      file: inputRef.current.files[0]
    });

    upChunk.on("error", (err) => {
      setErrorMessage(err.detail);
    });

    upChunk.on("progress", (progress) => {
      setProgress(Math.floor(progress.detail));
    });

    upChunk.on("success", () => {
      setIsPreparing(true);
    });
  };

```

`/pages/api/upload`
```js
export default async function uploadHandler(req, res) {
    const { method } = req;

    switch (method) {
        case "POST":
            try {
                const upload = await Video.Uploads.create({
                    new_asset_settings: { playback_policy: "public" },
                    cors_origin: "*"
                });
                res.json({
                    id: upload.id,
                    url: upload.url
                });
            } catch (e) {
                console.error("Request error", e);
                res.status(500).json({ error: "Error creating upload" });
            }
            break;
        default:
            res.setHeader("Allow", ["POST"]);
            res.status(405).end(`Method ${method} Not Allowed`);
    }
}

```

**Comment by icyJoseph** [comment]
I think you need to set an OPTIONS method handler, and that one should tell the browser the Access Control Allow Origin values. I am not sure though... - a bit rusty right now after long vacation.

**Reply by EhanDuan** [reply]
Thanks, I will check on OPTIONS method handler.
